### PR TITLE
[MM-54799] Convert ./components/widgets/menu/menu_header.tsx from Class Component to Function Component

### LINKED ISSUE #24763: Convert ./components/widgets/menu/menu_header.tsx from Class Component to Function Component
We want to move most of our codebase from Class Components to Function Components. This decision is based on [React recommendations](https://react.dev/reference/react/Component).

The task is to update the component to be a functional component, without any difference in functionality. Also, all related tests that fail may need refactoring.

Here are a few examples of the required changes:
https://github.com/mattermost/mattermost/pull/24187
https://github.com/mattermost/mattermost/pull/24210
https://github.com/mattermost/mattermost/pull/24183
https://github.com/mattermost/mattermost/pull/24181

If the tests do not work because they rely on the behavior of Class Components they also have to be migrated to functionaly test the same behavior. We prefer the use of `react-testing-library` for new tests, but it is not required.

When submitting the PR, please make sure to identify where and when the component shows, for QA to properly test the changes.

----

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA LINK: MM-54799

### PR DESCRIPTION
**Pull request by dipandhali2021**
#### Summary
This PR converts the MenuHeader component from class based component to functional based component

#### Ticket Link
Fixes #24763

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/components/widgets/menu/menu_header.tsx [hunk 1]
```diff
@@ -11,20 +10,14 @@ type Props = {
     onClick?: () => void;
 }
 
-/**
- * @deprecated Use the "webapp/channels/src/components/menu" instead.
- */
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/menu/menu_header.tsx:16]: please bring the deprecated notice

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/widgets/menu/menu_header.tsx (hunk 1 above)
Comment (dipandhali2021) [reply webapp/channels/src/components/widgets/menu/menu_header.tsx:16]: Due to this deprecated notice I am not able to to export the menuheader component at last that why I had to remove it. Please suggest the solution for it. #

Comment (larkox) [reply webapp/channels/src/components/widgets/menu/menu_header.tsx:16]: Try to put the deprecated notice just before the export.

File: webapp/channels/src/components/widgets/menu/menu_header.tsx [hunk 2]
```diff
@@ -11,20 +10,19 @@ type Props = {
...
+        <li
+            className='MenuHeader'
+            onClick={onClick}
+        >
+            {children}
+        </li>
+    );
+};
+
 /**
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/menu/menu_header.tsx:24]: This jsdoc comment should be above the component

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/widgets/menu/menu_header.tsx (hunk 2 above)
Comment (dipandhali2021) [reply webapp/channels/src/components/widgets/menu/menu_header.tsx:24]: ##Screenshots
before Component
Screenshot 2023-10-10 172124

after component 
Screenshot 2023-10-10 172152

Before placing the deprecated notice,error is coming.
hey @M-ZubairAhmed ,please have a look through it.

Comment (dipandhali2021) [reply webapp/channels/src/components/widgets/menu/menu_header.tsx:24]: Hello @M-ZubairAhmed,
Could you please provide a solution for the issue at hand? Your assistance would be helpful.

File: webapp/channels/src/components/widgets/menu/menu_header.tsx [hunk 3]
```diff
@@ -2,7 +2,6 @@
 // See LICENSE.txt for license information.
 
 import React from 'react';
-
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/menu/menu_header.tsx:5]: nit : Lets revert this change

File: webapp/channels/src/components/widgets/menu/menu_header.tsx [hunk 4]
```diff
@@ -14,17 +13,16 @@ type Props = {
...
-            <li
-                className='MenuHeader'
-                onClick={onClick}
-            >
-                {children}
-            </li>
-        );
-    }
-}
+const MenuHeader = ({children, onClick}: Props) => {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/menu/menu_header.tsx:17]: Can you please remove empty line between the component and the jsdoc

File: webapp/channels/src/components/widgets/menu/menu_header.tsx [hunk 5]
```diff
@@ -2,7 +2,6 @@
 // See LICENSE.txt for license information.
 
 import React from 'react';
-
 import './menu_header.scss';
 
 type Props = {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/menu/menu_header.tsx:7]: I see `divider` prop is not being used, can you remove that as well

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Review by M-ZubairAhmed — APPROVED** [review]
Thank you for working on this. I hope you had a good experience 😄

**Review by larkox — APPROVED** [review]
Looks good to me. Thank you very much!

**Comment by Tahanima** [comment]
Hi @lindy65, I would like to test this PR as a QA. Looking forward to your guidance.

**Comment by lindy65** [comment]
Hi @Tahanima - thanks for picking up QA on this PR 🎉 

Please test functionality around Custom Status (setting a custom status, clearing status, adding emoji icon, etc) accessible from Profile icon top right of the screen in webapp
 
image

image

Let me know if you have any questions!

**Comment by Tahanima** [comment]
`This week` option as a status expiration option does not work properly.

### Steps to reproduce
- Login to the test server using the credentials (Username: sysadmin, Password: Sys@dmin123)
- Now, click on the user avatar at the top right corner.
- Then, click on the 'Set a Custom Status' option.
- From the 'SUGGESTIONS' section, click on the `On a vacation This week` option and then press the 'Set Status' button.
- Now, hover over the user avatar. <ins>It shows `On a vacation Until Today`.</ins>
- Next, click on the user avatar. <ins>The custom status section shows `On a vacation (Until Today)`.</ins>
- Then, click on the custom status section. <ins>The 'Clear after:' shows `Today`. </ins>

### Operating System
Ubuntu 22.04.3 LTS

### Browser
Version 118.0.5993.88

### Recording

[bug_this_week_chrome.webm](https://github.com/mattermost/mattermost/assets/6233068/3b38c827-79a6-4f18-bf5f-188514db411c)

cc: @lindy65

**Comment by Tahanima** [comment]
I've tested the changes in the test server and it looks good to me.

### Steps
- Log in to the web application using the given credentials (Username: sysadmin, Password: Sys@dmin123)
- Click on the user avatar and set a custom status.
  - Choose status from a list of suggestions
  - Change expiration time
  - Select emoji
  - Create and select a custom emoji
  - Delete a custom emoji
  - Clear current status
  - Clear recent status
- Hover on the emoji icon and click on it after setting a custom status to open the 'Set a status' modal.

### Operating System
Ubuntu 22.04.3 LTS

### Browser
Version 118.0.5993.88

### Recording

[chrome.webm](https://github.com/mattermost/mattermost/assets/6233068/13b3d5d8-9829-4275-9dfd-9f5982399f19)


cc: @lindy65

**Review by lindy65 — APPROVED** [review]
Thanks @dipandhali2021 - looks good 👍 

Thanks for QA'ing @Tahanima 🎉 

cc @larkox

**Comment by larkox** [comment]
Thank you for the contribution!
